drivers/ethernet/eth_xlnx_gem.c

```diff
@@ -0,0 +1,1679 @@
+ * - VLAN tags not considered.
+ * - Wake-on-LAN interrupt not supported.
+ * - Send function is not SMP-capable (due to single TX done semaphore).
+ * - Interrupt-driven PHY management not supported - polling only.
+ * - No explicit placement of the DMA memory area(s) in either a
+ *   specific memory section or at a fixed memory location yet. This
+ *   is not an issue as long as the controller is used in conjunction
+ *   with the Cortex-R5 QEMU target or an actual R5 running without the
+ *   MPU enabled.
+ * - No detailed error handling when evaluating the Interrupt Status,
+ *   RX Status and TX Status registers.
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <devicetree.h>
+#include <sys/__assert.h>
+
+#include <net/net_if.h>
+#include <net/ethernet.h>
+#include <ethernet/eth_stats.h>
+
+#include "eth_xlnx_gem_priv.h"
+
+#define LOG_MODULE_NAME eth_xlnx_gem
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+static int  eth_xlnx_gem_dev_init(const struct device *dev);
+static void eth_xlnx_gem_iface_init(struct net_if *iface);
+static void eth_xlnx_gem_isr(const struct device *dev);
+static int  eth_xlnx_gem_send(const struct device *dev, struct net_pkt *pkt);
+static int  eth_xlnx_gem_start_device(const struct device *dev);
+static int  eth_xlnx_gem_stop_device(const struct device *dev);
+static enum ethernet_hw_caps
+	eth_xlnx_gem_get_capabilities(const struct device *dev);
+#if defined(CONFIG_NET_STATISTICS_ETHERNET)
+static struct net_stats_eth *eth_xlnx_gem_stats(const struct device *dev);
+#endif
+
+static void eth_xlnx_gem_reset_hw(const struct device *dev);
+static void eth_xlnx_gem_configure_clocks(const struct device *dev);
+static void eth_xlnx_gem_set_initial_nwcfg(const struct device *dev);
+static void eth_xlnx_gem_set_nwcfg_link_speed(const struct device *dev);
+static void eth_xlnx_gem_set_mac_address(const struct device *dev);
+static void eth_xlnx_gem_set_initial_dmacr(const struct device *dev);
+static void eth_xlnx_gem_init_phy(const struct device *dev);
+static void eth_xlnx_gem_poll_phy(struct k_work *item);
+static void eth_xlnx_gem_configure_buffers(const struct device *dev);
+static void eth_xlnx_gem_rx_pending_work(struct k_work *item);
+static void eth_xlnx_gem_handle_rx_pending(const struct device *dev);
+static void eth_xlnx_gem_tx_done_work(struct k_work *item);
+static void eth_xlnx_gem_handle_tx_done(const struct device *dev);
+
+static const struct ethernet_api eth_xlnx_gem_apis = {
+	.iface_api.init   = eth_xlnx_gem_iface_init,
+	.get_capabilities = eth_xlnx_gem_get_capabilities,
+	.send		  = eth_xlnx_gem_send,
+	.start		  = eth_xlnx_gem_start_device,
+	.stop		  = eth_xlnx_gem_stop_device,
+#if defined(CONFIG_NET_STATISTICS_ETHERNET)
+	.get_stats	  = eth_xlnx_gem_stats,
+#endif
+};
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+
+/* IRQ configuration function for GEM0 */
+ETH_XLNX_GEM_CONFIG_IRQ_FUNC(0);
```

Comment:
You could declare a super macro that expands to all of these, i.e.
ETH_XLNX_GEN_DEV(0) which would expand to the 5 macros below